src/mame/sega/segaai.cpp

```diff
@@ -0,0 +1,751 @@
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_JOYSTICK_UP) PORT_8WAY
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_JOYSTICK_DOWN) PORT_8WAY
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_JOYSTICK_LEFT) PORT_8WAY
+	PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_JOYSTICK_RIGHT) PORT_8WAY
+	PORT_BIT(0x10, IP_ACTIVE_LOW, IPT_BUTTON2) PORT_NAME("PL")
+	PORT_BIT(0x20, IP_ACTIVE_LOW, IPT_BUTTON1) PORT_NAME("RL")
+	PORT_BIT(0xc0, IP_ACTIVE_LOW, IPT_UNUSED)
+
+	PORT_START("PORT5")
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_BUTTON3) PORT_NAME("Grey Button")
+	PORT_BIT(0xfe, IP_ACTIVE_LOW, IPT_UNUSED)
+
+	PORT_START("TOUCH")
+	PORT_BIT(0x01, IP_ACTIVE_HIGH, IPT_BUTTON4) PORT_NAME("Press touchpad")
+
+	PORT_START("TOUCHPADX")
+	PORT_BIT(0xff, 0x80, IPT_LIGHTGUN_X) PORT_CROSSHAIR(X, 1.0, 0.0, 0) PORT_SENSITIVITY(30) PORT_KEYDELTA(10) PORT_MINMAX(0, 255) PORT_PLAYER(1) PORT_NAME("Touchpad X")
+
+	PORT_START("TOUCHPADY")
+	PORT_BIT(0xff, 0x80, IPT_LIGHTGUN_Y) PORT_CROSSHAIR(Y, 1.0, 0.0, 0) PORT_SENSITIVITY(30) PORT_KEYDELTA(10) PORT_MINMAX(0, 255) PORT_PLAYER(1) PORT_NAME("Touchpad Y")
+INPUT_PORTS_END
+
+
+void segaai_state::update_irq_state()
+{
+	int state = CLEAR_LINE;
+
+	if (m_irq_active & m_irq_enabled)
+	{
+		state = ASSERT_LINE;
+	}
+
+	m_maincpu->set_input_line(0, state);
+}
+
+
+// Based on edge triggers level triggers are created
+void segaai_state::vdp_interrupt(int state)
+{
+	if (state != CLEAR_LINE)
+	{
+		if (m_prev_v9938_irq == CLEAR_LINE)
+		{
+			m_irq_active |= IRQ_V9938;
+		}
+	}
+	m_prev_v9938_irq = state;
+
+	update_irq_state();
+}
+
+
+// Based on edge triggers level triggers are created
+void segaai_state::upd7759_drq_w(int state)
+{
+	int upd7759_irq = state ? CLEAR_LINE : ASSERT_LINE;
+	if (upd7759_irq != CLEAR_LINE)
+	{
+		if (m_prev_upd7759_irq == CLEAR_LINE)
+		{
+			m_irq_active |= IRQ_UPD7759;
+		}
+	}
+	m_prev_upd7759_irq = upd7759_irq;
+
+	update_irq_state();
+}
+
+
+IRQ_CALLBACK_MEMBER(segaai_state::irq_callback)
+{
+	if (m_irq_active & m_irq_enabled & IRQ_V9938)
+	{
+		m_vector = VECTOR_V9938;
+		m_irq_active &= ~IRQ_V9938;
+	}
+	else if (m_irq_active & m_irq_enabled & IRQ_UPD7759)
+	{
+		m_vector = VECTOR_UPD7759;
+		m_irq_active &= ~IRQ_UPD7759;
+	}
+	else
+	{
+		if (m_irq_active & m_irq_enabled)
+		{
+			fatalerror("Unknown irq triggered: $%02X active, $%02X enabled\n", m_irq_active, m_irq_enabled);
+		}
+		fatalerror("irq_callback called but no irq active and enabled: $%02X active, $%02X enabled\n", m_irq_active, m_irq_enabled);
+	}
+
+	update_irq_state();
+	return m_vector;
+}
+
+
+/*
+Mainboard 8255 port A
+
+ 76543210
+ +-------- Microphone sensor (1 = sound enabled)
+  +------- -BUSY output from the uPD7759
+   +------ PR trigger (active low)
+    +----- PL trigger (active low)
+     +---- Pad right (active low)
+      +--- Pad lefta (active low)
+       +-- Pad down (active low)
+        +- Pad up (active low)
+*/
+u8 segaai_state::i8255_porta_r()
+{
+	u8 data = (m_upd7759->busy_r() ? UPD7759_BUSY : 0) | (m_port4->read() & ~UPD7759_BUSY);
+
+	return data;
+}
```

Comment:
You can use `PORT_READ_LINE_DEVICE_MEMBER` to get the µPD7759 busy line state directly through the I/O port and avoid having this helper function.